common/containers/container.go

```diff
@@ -0,0 +1,224 @@
+// - Qualified names must be dot-delimited, e.g. `package.subpkg.name`.
+// - The last element in the qualified name is the simple name used as an alias.
+// - Alias names must not collide with each other.
+// - The alias name must not collide with the root-level container name.
+//
+// Aliases are distinct from container-based references in the following important ways:
+// - Containers follow C++ namespace resolution rules with searches from the most qualified name
+//   to the least qualified name.
+// - Container references within the CEL program may be relative, and are resolved to fully
+//   qualified names at either type-check time or program plan time, whichever comes first.
+// - Alias simple names must resolve to a fully-qualified name.
+// - Resolved aliases do not participate in namespace resolution.
+// - Resolved aliases are searched after container names, including container names in the global
+//   scope.
+//
+// If there is ever a case where an identifier could be in both the container and in the alias,
+// the container wins as the container will continue to evolve over time and the program must be
+// forward compatible with changes in the container.
+func Aliases(qualifiedNames ...string) ContainerOption {
+	return func(c *Container) (*Container, error) {
+		for _, qn := range qualifiedNames {
+			ind := strings.LastIndex(qn, ".")
+			alias := qn
+			if ind > 0 || ind < len(qn)-1 {
+				alias = qn[ind+1:]
+			}
+			_, err := AliasAs(qn, alias)(c)
+			if err != nil {
+				return nil, err
+			}
+		}
+		return c, nil
+	}
+}
+
+// AliasAs associates a fully-qualified name with a user-defined alias.
+//
+// In general, Aliases is preferred to AliasAs since the names generated from the Aliases option
+// are more easily traced back to source code. The AliasAs option is useful for propagating alias
+// configuration from one Container instance to another, and may also be useful for remapping
+// poorly chosen protobuf message / package names.
+//
+// Note: all of the rules that apply to Aliases also apply to AliasAs.
+func AliasAs(qualifiedName, alias string) ContainerOption {
+	return func(c *Container) (*Container, error) {
+		if len(alias) <= 0 || strings.Contains(alias, ".") {
+			return nil, fmt.Errorf(
+				"alias names must non-empty and simple (not qualified): alias=%s", alias)
+		}
+		ind := strings.LastIndex(qualifiedName, ".")
+		if ind <= 0 || ind == len(qualifiedName)-1 {
+			return nil, fmt.Errorf("aliases must refer to qualified names: %s",
+				qualifiedName)
+		}
+		aliasRef, found := c.aliases[alias]
+		if found {
+			return nil, fmt.Errorf(
+				"alias collides with existing reference: name=%s, alias=%s, existing=%s",
+				qualifiedName, alias, aliasRef)
+		}
+		if strings.HasPrefix(c.Name(), alias+".") || c.Name() == alias {
+			return nil, fmt.Errorf(
+				"alias collides with container name: name=%s, alias=%s, container=%s",
+				qualifiedName, alias, c.Name())
+		}
+		c.aliases[alias] = qualifiedName
```

Comment:
:bug: Can we make this work with a nil `c.aliases` so that `&Container{}` doesn’t panic? (Also, this means that `NewContainer` won’t have to make the map if there are no aliases.)
```suggestion
		if c.aliases == nil {
			c.aliases = make(map[string]string)
		}
		c.aliases[alias] = qualifiedName
```